Hide Outlines for 3D Models

srothst1 opened the pull request:
Fixes #8959

Added a new property `ignoreOutline` to Model.js and Cesium3DTileset.js. Based on this property, the line `ModelOutlineLoader.outlinePrimitives(this);` in Model.js will be either included or omitted. This exposes the ability to turn off outlines on 3D models.

The current issue that I am facing is that `this.ignoreOutline` is not being updated correctly in `Model.js`. I created a [sandcastle demo](http://localhost:8080/Apps/Sandcastle/index.html#c=hVLBbtswDP0VwicHcOWk61Y0TYNh2XZqkWILtosvikzHWmUpoGhnWZF/n2wnRdJl6E16eu9RfGSaLkrtwWiLQFi5Bj1wieBqbrFxZtN0hl7XlXhwOZp5j987mSOJUvovvxmt187CHRS1VRyO8QCegxvXZKGQxuMt7DKb2UYSNBo3SIFscQN75x8dFmeR6u4zZ1mGIpRFCTxnFoCRKCCP5Bod6o4PQkUoGX86Mvmip8SDJLO7we2hnPPVp1qbXNuVX2iDHjnUPpHPjyhxV24p1dNXqXBWm9DsCsZ9FwmkaeEodFYgoVUB4Da8RpoaIRyW2LLrdR5cc1COCBWbLWgLXXjil2/tfek2wZOpftvyTTO9so5wP5Yj11Mb6/jV705sDoF1cxBeoUWxJl1p1mEhhMzz+EyQZ0RKVkhSFGa7cH2WOXrWVrZr8TK2mSQOJ2nfiYJc9RlXhOjji+srMRzdJHA1FB9uRpcJXL8ftvMEcKQxLEXv0vkClCjb37y4PkguBbtvAZXWx5d7KcBasyr/R7s48Hb94kCURBPPW4PTXg3wUVdrRww1mViIlLFam5ChT5e1ekIWyvs2ipY6SY+lk1w3oPO7M4sNykjvw0sRduy7/oNZNJ2kgf+P1LiuzXmDZOS2pZWj6X0PCiEmabieV7JzZinplfNf) that showcases this issue. Please let me know what your suggestions are for resolving this.

Lastly, there was a discussion in #8959 about where this new property should be implemented. It is unclear if this feature should only be implemented for Cesium OSM Buildings. Feedback on this matter would also be apricated. 

@ebogo1 
@lilleyse

cesium-concierge commented:
Thanks for the pull request @srothst1!
* :heavy_check_mark: Signed CLA found.
* :grey_question: Unit tests were not updated.
    * Make sure you've [updated tests](https://github.com/CesiumGS/cesium/tree/master/Documentation/Contributors/TestingGuide) to reflect your changes, added tests for any new code, and ran the code coverage tool.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

ebogo1 commented on Source/Scene/Model.js:
```diff
@@ -5317,7 +5328,12 @@ Model.prototype.update = function (frameState) {
-        ModelOutlineLoader.outlinePrimitives(this);
+        console.log("Back Face Culling: " + this.backFaceCulling); //remove after testing in sandcastle
+        console.log("Show: " + this.show); //remove after testing in sandcastle
+        console.log("Ignore Outline: " + this.ignoreOutline); //remove after testing in sandcastle
```
I had to add the `ignoreOutline` option to `Batched3DModel3DTileContent.initialize` where it calls the Model constructor. In your Sandcastle, the debugger call stack showed that `Model` was being called from there so I took a look at the function. There are other places where a Model is created with a tileset's options, i.e. `Gltf3DTileContent` - maybe if this is only a change we want to apply to OSM Buildings then we don't need to update the option anywhere else?

srothst1 replied:
I think that this change should only apply to OSM Buildings. I just updated `Gltf3DTileContent` accordingly. However, I am open to feedback from our team and the community on how broad the ignore outlines functionality should be.

ebogo1 reviewed (commented):
These changes look mostly good to me. Do we want to allow users to swap this option after initialization?

srothst1 commented:
@ebogo1 It does not seem necessary to allow users to change this option after initialization. I have not seen anything in the community forum that suggests that this functionality is critical.

lilleyse commented on CHANGES.md:
```diff
@@ -10,6 +10,7 @@
 - Added checks for supported 3D Tiles extensions. [#9552](https://github.com/CesiumGS/cesium/issues/9552)
 - Added documentation clarifying that the `outlineWidth` property will be ignored on all major browsers on Windows platforms. [#9600](https://github.com/CesiumGS/cesium/pull/9600)
 - Added documentation for `KmlTour`, `KmlTourFlyTo`, and `KmlTourWait`. Added documentation and a `kmlTours` getter to `KmlDataSource`. Removed references to `KmlTourSoundCues`. [#8073](https://github.com/CesiumGS/cesium/issues/8073)
+- Added the ability for users to turn of outlines for 3D models. [#8959](https://github.com/CesiumGS/cesium/issues/8959)
```
I think OSM should be highlighted here since it's the main thing using [CESIUM_primitive_outline](https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Vendor/CESIUM_primitive_outline).

"Added the ability to hide outlines on OSM Buildings and other tilesets and glTF models using the [CESIUM_primitive_outline](https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Vendor/CESIUM_primitive_outline) extension. [#8959](https://github.com/CesiumGS/cesium/issues/8959)"

lilleyse commented on Source/Scene/Cesium3DTileset.js:
```diff
@@ -96,6 +96,7 @@ import TileOrientedBoundingBox from "./TileOrientedBoundingBox.js";
  * @param {Cartesian3[]} [options.sphericalHarmonicCoefficients] The third order spherical harmonic coefficients used for the diffuse color of image-based lighting.
  * @param {String} [options.specularEnvironmentMaps] A URL to a KTX file that contains a cube map of the specular lighting and the convoluted specular mipmaps.
  * @param {Boolean} [options.backFaceCulling=true] Whether to cull back-facing geometry. When true, back face culling is determined by the glTF material's doubleSided property; when false, back face culling is disabled.
+ * @param {Boolean} [options.ignoreOutline=false] Whether to ignore the 3D model outline. When false, outlines are displayed. When true, all outlines are not displayed.
```
The description should also mention that this applies to models using the [CESIUM_primitive_outline](https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Vendor/CESIUM_primitive_outline) extension

lilleyse commented on Source/Scene/Model.js:
```diff
@@ -5317,7 +5328,9 @@ Model.prototype.update = function (frameState) {
         loadResources.pendingShaderLoads === 0
       ) {
-        ModelOutlineLoader.outlinePrimitives(this);
+        if (this.ignoreOutline === false) {
```
More concise:

```suggestion
        if (!this.ignoreOutline) {
```

lilleyse commented on Source/Scene/Cesium3DTileset.js:
```diff
@@ -96,6 +96,7 @@ import TileOrientedBoundingBox from "./TileOrientedBoundingBox.js";
  * @param {Cartesian3[]} [options.sphericalHarmonicCoefficients] The third order spherical harmonic coefficients used for the diffuse color of image-based lighting.
  * @param {String} [options.specularEnvironmentMaps] A URL to a KTX file that contains a cube map of the specular lighting and the convoluted specular mipmaps.
  * @param {Boolean} [options.backFaceCulling=true] Whether to cull back-facing geometry. When true, back face culling is determined by the glTF material's doubleSided property; when false, back face culling is disabled.
+ * @param {Boolean} [options.ignoreOutline=false] Whether to ignore the 3D model outline. When false, outlines are displayed. When true, all outlines are not displayed.
```
This option should also be documented in `createOsmBuildings`

lilleyse commented on Source/Scene/Cesium3DTileset.js:
```diff
@@ -776,6 +777,16 @@ function Cesium3DTileset(options) {
    */
   this.backFaceCulling = defaultValue(options.backFaceCulling, true);
 
+  /**
+   * Whether to ignore the 3D model outline. When true, all outlines are not
+   * displayed. When false, outlines are displayed.
+   *
+   * @type {Boolean}
+   *
+   * @default false
+   */
+  this.ignoreOutline = defaultValue(options.ignoreOutline, false);
```
Should be marked as `@readonly` since it can't be changed after construction

lilleyse replied:
Same comment for `Model`.

lilleyse commented:
> @ebogo1 It does not seem necessary to allow users to change this option after initialization. I have not seen anything in the community forum that suggests that this functionality is critical.

Ideally it would be toggleable. However I don't think the current Model code is flexible enough for that right now, and it's not a dealbreaker for this PR.

CC @ptrgags @sanjeetsuhag when you go to integrate `CESIUM_primitive_outline` in the new model code try to make it flexible enough for toggling on/off after construction.

lilleyse commented on Source/Scene/Cesium3DTileset.js:
```diff
@@ -776,6 +777,17 @@ function Cesium3DTileset(options) {
    */
   this.backFaceCulling = defaultValue(options.backFaceCulling, true);
 
+  /**
+   * Whether to ignore the 3D model outline. When true, all outlines are not
+   * displayed. When false, outlines are displayed.
+   *
+   * @type {Boolean}
+   * @readonly
+   *
+   * @default false
+   */
+  this.showOutline = defaultValue(options.showOutline, true);
```
Make sure the update the documentation - both the description and the `@default`

lilleyse replied:
Same comment for the `Model` documentation

lilleyse commented on Source/Scene/ModelInstanceCollection.js:
```diff
@@ -164,6 +165,8 @@ function ModelInstanceCollection(options) {
   this.backFaceCulling = defaultValue(options.backFaceCulling, true);
   this._backFaceCulling = this.backFaceCulling;
+  this.showOutline = defaultValue(options.showOutline, true);
+  this._showOutline = this.showOutline;
```
`showOutline` needs to be passed through when the model is constructed. See [`modelOptions`](https://github.com/CesiumGS/cesium/blob/4fb9fbb097663898733ef5850f4df7d0174f3af0/Source/Scene/ModelInstanceCollection.js#L673-L697) - it should be added there

lilleyse commented on Source/Scene/createOsmBuildings.js:
```diff
@@ -21,6 +21,8 @@ import Cesium3DTileStyle from "./Cesium3DTileStyle.js";
  *        specified, a default style is used which gives each building or building part a
  *        color inferred from its OpenStreetMap <code>tags</code>. If no color can be inferred,
  *        <code>options.defaultColor</code> is used.
+ * @param {Boolean} [options.showOutline=true] Whether to display the outline for models. When true, outlines \
+ *        are displayed. When false, outlines are not displayed.
```
Remove the `\` as it will show up in the generated documentation.

![Selection_029](https://user-images.githubusercontent.com/915398/123323472-8a7d0780-d503-11eb-9b8d-efd4d52fd2ef.png)

lilleyse replied:
Maybe a slightly more tailored description for `createOsmBuildings`:  "Whether to show outlines around buildings..."

lilleyse reviewed (commented):
Getting close...

srothst1 commented:
@lilleyse I just
1. updated the documentation for `Cesium3DTileset.js` and `Model.js`
2. changed where `showOutline` is added in `ModelInstanceCollection.js`
3. removed the `\` from `createOsmBuildings.js`
4. added a more tailored description of `showOutline` to `createOsmBuildings.js`

Hopefully I didn't miss anything - let me know!

lilleyse commented on Source/Scene/Model.js:
```diff
@@ -5317,7 +5330,9 @@ Model.prototype.update = function (frameState) {
         loadResources.pendingShaderLoads === 0
       ) {
-        ModelOutlineLoader.outlinePrimitives(this);
+        if (this.showOutline) {
+          ModelOutlineLoader.outlinePrimitives(this);
+        }
```
Could you add a test for this in `ModelSpec.js`? I think it can be as simple as loading any model with `showOutline: false` and checking that `ModelOutlineLoader.outlinePrimitives` is not called (using a spy).

lilleyse commented:
The code looks good! I just left one comment about testing.

srothst1 commented:
@lilleyse I just added a spec to `ModelSpec.js`. This spec should test loading a glTF 2.0 model with `showOutline : false` and ensuring that `ModelOutlineLoader.outlinePrimitives` is not called by using a spy `spyOn(ModelOutlineLoader, "outlinePrimitives");`.

lilleyse commented on Specs/Scene/ModelSpec.js:
```diff
@@ -1206,6 +1207,26 @@ describe(
       });
     });
 
+    it("loads a glTF 2.0 model with showOutline set as false", function () {
+      return loadModel(boxGltf2Url).then(function (m) {
+        verifyRender(m);
+        m.show = true;
+        m.luminanceAtZenith = undefined;
+        m.showOutline = false;
+        spyOn(ModelOutlineLoader, "outlinePrimitives");
+
+        expect({
+          scene: scene,
+          time: JulianDate.fromDate(new Date("January 1, 2014 12:00:00 UTC")),
+        }).toRenderAndCall(function (rgba) {
+          expect(rgba).toEqualEpsilon([179, 9, 9, 255], 5); // Red
+        });
+        expect(ModelOutlineLoader.outlinePrimitives).not.toHaveBeenCalled();
+
+        primitives.remove(m);
+      });
+    });
```
The test can be simplified by using the `options` parameter for `loadModel`

```js
    it("loads a glTF 2.0 model with showOutline set as false", function () {
      spyOn(ModelOutlineLoader, "outlinePrimitives");
      return loadModel(boxGltf2Url, {
        showOutline: false
      }).then(function (m) {
        expect(ModelOutlineLoader.outlinePrimitives).not.toHaveBeenCalled();
        primitives.remove(m);
      });
    });
```

srothst1 commented:
@lilleyse just committed your suggestions.

lilleyse commented:
Thanks @srothst1!
